crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -243,105 +248,117 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
 
         let storage_changeset =
             self.get_or_take::<tables::StorageChangeSet, UNWIND>(storage_range)?;
-        let account_changeset = self.get_or_take::<tables::AccountChangeSet, UNWIND>(range)?;
+        let account_changeset =
+            self.get_or_take::<tables::AccountChangeSet, UNWIND>(range.clone())?;
 
         // iterate previous value and get plain state value to create changeset
         // Double option around Account represent if Account state is know (first option) and
         // account is removed (Second Option)
 
-        let mut state: BundleStateInit = HashMap::new();
-
+        let mut bundle_builder = BundleBuilder::new(range);
         // This is not working for blocks that are not at tip. as plain state is not the last
         // state of end range. We should rename the functions or add support to access
         // History state. Accessing history state can be tricky but we are not gaining
         // anything.
         let mut plain_accounts_cursor = self.tx.cursor_write::<tables::PlainAccountState>()?;
         let mut plain_storage_cursor = self.tx.cursor_dup_write::<tables::PlainStorageState>()?;
 
-        let mut reverts: RevertsInit = HashMap::new();
+        // list of all seen addresses
+        let mut addresses_seen = vec![];
 
         // add account changeset changes
         for (block_number, account_before) in account_changeset.into_iter().rev() {
             let AccountBeforeTx { info: old_info, address } = account_before;
-            match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((old_info, new_info, HashMap::new()));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    // overwrite old account state.
-                    entry.get_mut().0 = old_info;
-                }
+            if !addresses_seen.contains(&address) {
+                // save address as seen
+                addresses_seen.push(address);
+                let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
+                bundle_builder = bundle_builder.state_present_account_info(
+                    address,
+                    into_revm_acc(new_info.unwrap_or_default()),
```

Comment:
This line does not look good. Having `Some` and having `AccountInfo::default` are two different account states. Should probably be handled differently

Reply:
Looked at the BundleBuilder and in this case it would create `Some(AccountInfo::default())` and I assume we would expect `None`:
https://github.com/bluealloy/revm/blob/main/crates/revm/src/db/states/bundle_state.rs#L166-L167

Reply:
yes makes sense. I think it should work if first checking if `new_info` (or `old_info` and `present_info` later in the code) is `Some` and only if it is so, then do `bundle_builder.state_present_account_info()`.